pkg/algorithm/cell_allocation.go

```diff
@@ -168,45 +261,54 @@ func mapPhysicalCellToVirtual(
 		if parentVirtual == nil {
 			return nil, message
 		} else {
-			return getLowestPriorityCell(parentVirtual.GetChildren(), p).(*VirtualCell), ""
+			return getLowestPriorityVirtualCell(parentVirtual.GetChildren(), p), ""
 		}
 	}
 }
 
-// getLowestPriorityCell returns a cell with the lowest priority among the cells
-// whose priorities are lower than the given priority (p).
-func getLowestPriorityCell(cl CellList, p CellPriority) Cell {
+// getLowestPriorityVirtualCell returns a virtual cell with the lowest priority
+// among those whose priorities are lower than the given priority (p).
+// We don't just return a free cell because in cases like reconfiguration, there might be no
+// free cell left. Then we allow to return a non-free, but lowest-priority cell and preempt it.
+func getLowestPriorityVirtualCell(cl CellList, p CellPriority) (lowestPriorityCell *VirtualCell) {
 	lowestPriority := maxGuaranteedPriority
-	var lowestPriorityCell Cell
 	for _, c := range cl {
-		pp := c.GetPriority()
-		if pp == freePriority {
-			return c
-		} else if pp < p && pp < lowestPriority {
-			lowestPriority = pp
-			lowestPriorityCell = c
+		vc := c.(*VirtualCell)
+		priority := vc.GetPriority()
+		if priority == freePriority {
+			if vc.GetPhysicalCell() == nil {
+				return vc
+			} else {
+				// Although freePriority must be the lowest priority, we should not return a free cell
+				// with binding, because such binding cannot be preempted (e.g., the binding is
```

Comment:
Comment here
```
freePriority          = opportunisticPriority - 1
```
that free cell may also have binding such as it is a (doomed) bad cell?